bin/util/transifex.js

```diff
@@ -0,0 +1,574 @@
+  json,
+  (_, value) => {
+    if (value != null && typeof value === 'object' &&
+      typeof value.string === 'string')
+      return PluralForms.fromTransifex(value.string);
+    return value;
+  }
+);
+
+
+
+////////////////////////////////////////////////////////////////////////////////
+// READ SOURCE MESSAGES
+
+// Returns the Vue I18n messages for the source locale after converting them to
+// PluralForms objects.
+const readSourceMessages = (localesDir, filenamesByComponent) => {
+  // Read the root messages.
+  const reviver = (key, value) => {
+    if (typeof value === 'string') return PluralForms.fromVueI18n(value);
+    if (key === 'full') {
+      // `value` will be an array if $tcPath() is used.
+      if (!Array.isArray(value)) logThenThrow(value, 'invalid full property');
+      return new PluralForms(value.map(pluralForms => {
+        if (pluralForms.length !== 1)
+          logThenThrow(value, 'invalid full property');
+        return pluralForms[0];
+      }));
+    }
+    return value;
+  };
+  const messages = parse(
+    fs.readFileSync(`${localesDir}/${sourceLocale}.json`).toString(),
+    reviver
+  );
+
+  // Read the component messages.
+  messages.component = {};
+  for (const [componentName, filename] of filenamesByComponent) {
+    const content = fs.readFileSync(filename).toString();
+    const match = content.match(/<i18n( +lang="json5")? *>/);
+    if (match != null) {
+      const begin = match.index + match[0].length;
+      const end = content.indexOf('</i18n>', begin);
+      if (end === -1) logThenThrow(filename, 'invalid single file component');
+      // Trimming so that if there is an error, the line number is clear.
+      const json = content.slice(begin, end).trim();
+      try {
+        messages.component[componentName] = parse(json, reviver)[sourceLocale];
+      } catch (e) {
+        // eslint-disable-next-line no-console
+        console.error(`could not parse the Vue I18n JSON of ${componentName}`);
+        throw e;
+      }
+    }
+  }
+
+  return messages;
+};
+
+
+
+////////////////////////////////////////////////////////////////////////////////
+// WRITE TRANSLATIONS
+
+// Stores a source message with the corresponding translation.
+class Translation {
+  constructor(parent, key) {
+    this.parent = parent;
+    this._key = key;
+    this.source = parent._source[key];
+  }
+
+  get root() { return this.parent.root; }
+
+  // The translation is "live": if it changes in the parent, it will change here
+  // as well.
+  get translated() { return this.parent._translated[this._key]; }
+
+  toJSON(key) {
+    if (this.translated.isEmpty()) return undefined;
+    return key === 'full' && this.translated.length !== 1
+      ? Array.from(this.translated)
+      : this.translated.toVueI18n();
+  }
+}
+
+// `Translations` stores an object of source messages with the corresponding
+// translations. It provides methods to modify the translations.
+class Translations {
```

Comment:
Comment: mirrors the Transifex json structure.

Reply:
Ooh, right, it mirrors the source Vue I18n json and reads the Transifex translations into it. This is a very helpful comment.